src/nets/e2e_asr_attctc_th.py

```diff
@@ -1897,6 +1967,212 @@ def recognize_beam(self, h, lpz, recog_args, char_list, rnnlm=None):
         # remove sos
         return nbest_hyps
 
+    def recognize_beam_batch(self, h, hlens, lpz, recog_args, char_list, rnnlm=None):
+        '''beam search implementation
+
+        :param Variable h:
+        :param Namespace recog_args:
+        :param char_list:
+        :return:
+        '''
+        logging.info('input lengths: ' + str(h.size(0)))
+        h = mask_by_length(h, hlens, 0)
+        
+        # search parms
+        batch = len(hlens)
+        beam = recog_args.beam_size
+        penalty = recog_args.penalty
+        ctc_weight = recog_args.ctc_weight
+        att_weight = 1.0 - ctc_weight
+
+        n_bb = batch * beam
+        n_bo = beam * self.odim
+        n_bbo = n_bb * self.odim
+
+        max_hlen = max(hlens)
+        if recog_args.maxlenratio == 0:
+            maxlen = max_hlen
+        else:
+            maxlen = max(1, int(recog_args.maxlenratio * max_hlen))
+        minlen = int(recog_args.minlenratio * max_hlen)
+        logging.info('max output length: ' + str(maxlen))
+        logging.info('min output length: ' + str(minlen))
+
+        # initialization
+        c_prev = [to_cuda(self, Variable(torch.FloatTensor(n_bb, self.dunits).zero_(),
+                                         volatile=True))
+                  for _ in range(self.dlayers)]
+        z_prev = [to_cuda(self, Variable(torch.FloatTensor(n_bb, self.dunits).zero_(),
+                                         volatile=True))
+                  for _ in range(self.dlayers)]
+        c_list = [to_cuda(self, Variable(torch.FloatTensor(n_bb, self.dunits).zero_(),
+                                         volatile=True))
+                  for _ in range(self.dlayers)]
+        z_list = [to_cuda(self, Variable(torch.FloatTensor(n_bb, self.dunits).zero_(),
+                                         volatile=True))
+                  for _ in range(self.dlayers)]
+        vscores = to_cuda(self, Variable(torch.FloatTensor(batch, beam).zero_(),
+                                         volatile=True))
+        a_prev = None
+        rnnlm_prev = None
+
+        # reset pre-computation of h
+        self.att.reset()
+
+        # preprate sos
+        y = self.sos
+        yseq = [[y] for _ in range(n_bb)]
+        stop_search = [False for _ in range(batch)]
+        nbest_hyps = [[] for _ in range(batch)]
+
+        if lpz is not None:
+            ctc_prefix_score = BatchCTCPrefixScoreTH(lpz, 0, self.eos, beam)
+            ctc_states_prev = ctc_prefix_score.initial_state()
+            ctc_scores_prev = to_cuda(self, torch.Tensor(batch, n_bo).zero_())
+
+        # expansion
+        ended_hyps = [[] for _ in range(batch)]
+        exp_hlens = torch.LongTensor(hlens * beam).view(beam, batch).transpose(0, 1).contiguous()
```

Comment:
This does not work for me. So I changed to 
`exp_hlens = torch.LongTensor(hlens).repeat(beam).view(beam, batch).transpose(0, 1).contiguous()`
and it works.
Not sure if this is the best approach.
Thx